src/Stratis.Bitcoin.Features.MemoryPool/MempoolValidator.cs

```diff
@@ -438,6 +438,19 @@ private async Task AcceptToMemoryPoolWorkerAsync(MempoolValidationState state, T
                     return;
                 }
 
+                // Dont accept transactions to the memory pool that are already known on the valid chain.
+                FetchCoinsResponse fetchCoinsResponse = this.coinView.FetchCoins(new[] { tx.GetHash() });
```

Comment:
A trx that is already in a block will be rejected because its inputs are already spent, why do we need to make another check?

Reply:
Why would the inputs necessarily already been spent. FetchCoins returns a made-up default value which make it difficult to know if this is something we've seen before.

Note that this code also gets called when loading the mempool from persistence. So that would be one way for the tx to sneak back into the mempool.

Reply:
its the same thing is what I am saying

Reply:
@dangershony it might be a valid use case to catch these earlier? So we don't go through the process of first validating the transaction only then to found out that it is already on the chain?

Reply:
@fassadlr bear in mind that we need to balance the overhead of performing such validation. If peers can trivially cause us to have to expend resources then a DoS vector results. It would be interesting to compare the processing required to validate signatures etc. vs the lookup from the coinview (which in the worst case might hit the disk).